pkg/lexer/lexer.go

```diff
@@ -0,0 +1,150 @@
+// Copyright © 2021 Rak Laptudirm <raklaptudirm@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lexer
+
+import (
+	"fmt"
+	"unicode/utf8"
+
+	"github.com/raklaptudirm/mash/pkg/token"
+)
+
+var (
+	ErrIllegalNUL = fmt.Errorf("illegal character NUL")
+	ErrIllegalBOM = fmt.Errorf("illegal byte order mark")
+	ErrIllegalEnc = fmt.Errorf("illegal utf-8 encoding")
+)
+
+type lexer struct {
+	src  string
+	ch   rune
+	prev token.TokenType
+
+	Tokens chan token.Token
+
+	err ErrorHandler
+
+	offset   int
+	rdOffset int
+
+	start token.Position
+	pos   token.Position
+
+	ErrorCount int
+}
+
+const (
+	eof = -1     // end of file
+	bom = 0xFEFF // byte order mark
+)
+
+type ErrorHandler func(token.Position, error)
```

Comment:
Nice!

Reply:
I figured that the callers may require additional info or use those as real error values, so I figured it is just better to use the error interface.